skbio/diversity/_block.py

```diff
@@ -0,0 +1,303 @@
+        Block row IDs to keep in the counts matrix.
+    col_ids : 1D np.ndarray of int
+        Block column IDs to keep in the counts matrix. Note, these correspond
+        to rows in the counts matrix, but columns in a subsequent distance
+        matrix.
+    kwargs : dict
+        Keyword arguments containing information about the block to compute.
+
+    Returns
+    -------
+    dict
+        kwargs that describe the block to compute. A filtered ``counts`` matrix
+        is stored in kwargs. If applicable, a filtered ``tree`` and ``otu_ids``
+        are also stored.
+    """
+    # filter kwargs to valid keys for downstream block compute
+    # explicit copies are not necessary as implicit copies or overwwrites are
+    # performed below.
+    valid_block_keys = {'counts', 'ids', 'tree', 'otu_ids', 'metric',
+                        'id_pairs', 'validate'}
+    block_kwargs = {k: v for k, v in kwargs.items() if k in valid_block_keys}
+
+    ids_to_keep = np.unique(np.hstack([row_ids, col_ids]))
+
+    # create a view of the relevant samples
+    counts_block = counts[ids_to_keep]
+
+    # remove from the block any empty observations
+    # NOTE: this will perform an implicit copy
+    nonzero_cols = (counts_block != 0).any(axis=0)
+    counts_block = counts_block[:, nonzero_cols]
+
+    block_kwargs['counts'] = counts_block
+    block_kwargs['ids'] = ids_to_keep
+
+    if 'tree' in kwargs and 'otu_ids' in kwargs:
+        block_kwargs['otu_ids'] = np.asarray(kwargs['otu_ids'])[nonzero_cols]
+        block_kwargs['tree'] = kwargs['tree'].shear(block_kwargs['otu_ids'])
+
+    return block_kwargs
+
+
+def _pairs_to_compute(rids, cids):
+    """Determine the pairs of samples to compute distances between
+
+    Parameters
+    ----------
+    rids : Iterable
+        The row IDs in the partial pairwise computation.
+    cids : Iterable
+        The column IDs in the partial pairwise computation.
+
+    Raises
+    ------
+    ValueError
+        When determining ID pairs for blocks that fall outside of the diagonal
+        of the resulting distance matrix, if a pair corresponds to the lower
+        triangle, complain loudly.
+
+    Returns
+    -------
+    list of tuple
+        The ID pairs to compute distances between.
+    """
+    # if identical, gather the upper triangle
+    if len(rids) == len(cids) and (rids == cids).all():
+        return [(i, j) for idx, i in enumerate(rids) for j in rids[idx+1:]]
+
+    # otherwise, grab pairwise combinations disregarding the diagonal
+    else:
+        if set(rids).intersection(set(cids)):
+            raise ValueError("Attempting to compute a lower triangle")
+        return [(i, j) for i in rids for j in cids if i != j]
+
+
+def _block_kwargs(**kwargs):
+    """Construct arguments describing a block to compute
+
+    Returns
+    -------
+    dict
+        The parameters for the block of the distance matrix to compute.
+    """
+    for row_ids, col_ids in _generate_id_blocks(**kwargs):
```

Comment:
oh, I see now why you need `kwargs` in `_generate_id_blocks` func param. however, would it be more clear to just pass `kwargs['ids']` and `kwargs['k']` explicitly?